lib/wasi/src/state/mod.rs

```diff
@@ -1938,17 +345,43 @@ impl WasiState {
         });
         Ok(ret)
     }
+
+    /// Forking the WasiState is used when either fork or vfork is called
+    pub fn fork(&self, inc_refs: bool) -> Self {
+        WasiState {
+            fs: self.fs.fork(inc_refs),
+            secret: self.secret.clone(),
+            inodes: self.inodes.clone(),
+            threading: Default::default(),
+            futexs: Default::default(),
+            clock_offset: Mutex::new(self.clock_offset.lock().unwrap().clone()),
+            bus: Default::default(),
+            args: self.args.clone(),
+            envs: self.envs.clone(),
+            preopen: self.preopen.clone(),
+            runtime: self.runtime.clone(),
+        }
+    }
 }
 
-pub fn virtual_file_type_to_wasi_file_type(file_type: wasmer_vfs::FileType) -> Filetype {
-    // TODO: handle other file types
-    if file_type.is_dir() {
-        Filetype::Directory
-    } else if file_type.is_file() {
-        Filetype::RegularFile
-    } else if file_type.is_symlink() {
-        Filetype::SymbolicLink
-    } else {
-        Filetype::Unknown
+#[derive(Debug, Clone)]
+pub struct WasiDummyWaker;
```

Comment:
What is this? Can we have code comments on the use?

Reply:
Will add some comments.

In short: if you want to manually poll a future outside of an async runtime, you have to provide a "fake" no-op waker, because `poll()` requires a `Context` argument , which can only be constructed with a waker.
( I implemented the same thing at least twice before)

Reply:
Moved to utils submodule and added a doc comment in 9aa6a07e7ba139be580b5739b6321217484a9f55 .